bosh-director/lib/bosh/director/vm_deleter.rb

```diff
@@ -17,7 +18,11 @@ def delete_for_instance(instance)
 
     def delete_vm(vm_cid)
       @logger.info('Deleting VM')
-      @error_ignorer.with_force_check { @cloud.delete_vm(vm_cid) }
+      @error_ignorer.with_force_check {
+        unless @enable_virtual_delete_vm
+          @cloud.delete_vm(vm_cid)
```

Comment:
@cppforlife 

Not quite correct, the PR is including not only the changes of vm_deleter.rb but also instance_updater.rb. So there are 3 scenarios:
1. For deployment upgrade / instance update, if enable_virtual_delete_vm=true, bosh director will not invoke delete_vm action of CPI, that is what we want, because we can use os_reload to achieve instance upgrade.
2. For bosh delete deployment command, even if enable_virtual_delete_vm=true, the command will also delete vms.
3. For delete_vm resolution by bosh cck,  even if enable_virtual_delete_vm=true, the solution will delete that vm.
As a whole, the change will cause no impact to other CPI/infrastructure, such as, aws, openstack, vsphere. For Softlayer, if we want to enable the virtual delete action during instance upgrade, we will set enable_virtual_delete_vms = true in director.yml